### PR TITLE
Add `Language.distill`

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This method is the distillation counterpart of `Language.update`.  It takes a teacher `Language` instance and distills the student pipes on the teacher pipes.

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy/language.py [hunk 1]
```diff
@@ -1018,6 +1018,98 @@ def __call__(
...
+        sgd: Optional[Optimizer] = None,
+        losses: Optional[Dict[str, float]] = None,
+        component_cfg: Optional[Dict[str, Dict[str, Any]]] = None,
+        exclude: Iterable[str] = SimpleFrozenList(),
+        annotates: Iterable[str] = SimpleFrozenList(),
+        component_map: Optional[Dict[str, str]] = None,
+    ):
+        """Update the models in the pipeline.
+        teacher (Language): Teacher to distill from.
+        examples (Iterable[Example]): A batch of examples
```
Comment (shadeMe) [inline spacy/language.py:1036]: Could we expand the description here to indicate that `reference` pertains to the `Doc` that's passed to the teacher (and `predicted` to the student)?

File: website/docs/api/language.mdx [hunk 2]
```diff
@@ -333,6 +333,34 @@ and custom registered functions if needed. See the
 | `component_cfg` | Optional dictionary of keyword arguments for components, keyed by component names. Defaults to `None`. ~~Optional[Dict[str, Dict[str, Any]]]~~ |
 | **RETURNS**     | The updated `losses` dictionary. ~~Dict[str, float]~~                                                                                          |
 
+## Language.distill {id="distill",tag="method",version="4"}
```
Comment (shadeMe) [inline website/docs/api/language.mdx:336]:
```suggestion
## Language.distill {id="distill",tag="method,experimental",version="4"}
```
To keep it in line with [this PR](https://github.com/explosion/spaCy/pull/12016/files).

File: spacy/language.py [hunk 3]
```diff
@@ -1018,6 +1018,98 @@ def __call__(
...
+                # A missing teacher pipe is not an error, some student pipes
+                # do not need a teacher, such as tok2vec layer losses.
+                teacher_pipe_name = (
+                    component_map[name] if name in component_map else name
+                )
+                teacher_pipe = teacher_pipes.get(teacher_pipe_name, None)
+                student_proc.distill(
+                    teacher_pipe,
+                    examples,
+                    sgd=None,
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/language.py:1093]: Good question! I followed the same structure as the update method, but in both cases it seems ok to just pass the optimizer and let the pipe do the work. The primary reason I can think of is that this makes it easier to customize the optimization behavior. With the current setup, you only have to override the `finish_update` method of your pipe. Whereas if we pass the optimizer to `distill`/`update`, you'll have to reimplement the complete `distill`/`update` method.

File: spacy/language.py (hunk 1 above)
Comment (danieldk) [reply spacy/language.py]: Fixed. Also updated the docs for trainable pipes to mention this.

File: spacy/language.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/language.py]: Changed following internal discussions.

**Review by shadeMe — APPROVED** [review]

File: spacy/language.py [hunk 4]
```diff
@@ -1018,6 +1018,100 @@ def __call__(
...
+        *,
+        drop: float = 0.0,
+        sgd: Optional[Optimizer] = None,
+        losses: Optional[Dict[str, float]] = None,
+        component_cfg: Optional[Dict[str, Dict[str, Any]]] = None,
+        exclude: Iterable[str] = SimpleFrozenList(),
+        annotates: Iterable[str] = SimpleFrozenList(),
+        component_map: Optional[Dict[str, str]] = None,
+    ):
+        """Update the models in the pipeline.
```
Comment (svlandeg) [inline spacy/language.py:1034]: Maybe have something slightly more specific here (currently c/p from `update`)

File: spacy/language.py [hunk 5]
```diff
@@ -1018,6 +1018,100 @@ def __call__(
...
+        drop (float): The dropout rate.
+        sgd (Optional[Optimizer]): An optimizer.
+        losses (Optional(Dict[str, float])): Dictionary to update with the loss,
+            keyed by component.
+        component_cfg (Optional[Dict[str, Dict[str, Any]]]): Config parameters
+            for specific pipeline components, keyed by component name.
+        exclude (Iterable[str]): Names of components that shouldn't be updated.
+        annotates (Iterable[str]): Names of components that should set
+            annotations on the predicted examples after updating.
+        component_map (Optional[Dict[str, str]]): Map student pipe name to
```
Comment (svlandeg) [inline spacy/language.py:1048]: Should we name this something more specific, like `student_to_teacher`?

File: spacy/language.py [hunk 6]
```diff
@@ -1018,6 +1018,100 @@ def __call__(
...
+            component_cfg = {}
+        pipe_kwargs = {}
+        for name, student_proc in self.pipeline:
+            component_cfg.setdefault(name, {})
+            pipe_kwargs[name] = deepcopy(component_cfg[name])
+            component_cfg[name].setdefault("drop", drop)
+            pipe_kwargs[name].setdefault("batch_size", self.batch_size)
+
+        teacher_pipes = dict(teacher.pipeline)
+        for name, student_proc in self.pipeline:
```
Comment (svlandeg) [inline spacy/language.py:1080]: I propose to use `student_name` here. Though slightly more verbose, it'd make the code more readible, especially around L1102 where you're switching between the student/teacher names.

File: spacy/language.py [hunk 7]
```diff
@@ -1243,12 +1337,16 @@ def initialize(
...
+        labels: Optional[Dict[str, Any]] = None,
         sgd: Optional[Optimizer] = None,
     ) -> Optimizer:
         """Initialize the pipe for training, using data examples if available.
 
         get_examples (Callable[[], Iterable[Example]]): Optional function that
             returns gold-standard Example objects.
+        labels (Dict[str, Any]): Labels to pass to pipe initialization, using
+            the names of the pipes as keys. Overrides labels that are in the
+            model configuration.
```
Comment (svlandeg) [inline spacy/language.py:1349]:
```suggestion
        labels (Optional[Dict[str, Any]]): Labels to pass to pipe initialization, using
            the names of the pipes as keys. Overrides labels that are in the
            model configuration.
```

File: spacy/tests/test_language.py [hunk 8]
```diff
@@ -799,3 +805,65 @@ def bad_pipe(doc):
...
+        Example.from_dict(teacher.make_doc(t[0]), {}) for t in TAGGER_TRAIN_DATA
+    ]
+
+    component_map = (
+        None
+        if teacher_tagger.name == student_tagger.name
+        else {student_tagger.name: teacher_tagger.name}
+    )
+
+    for i in range(50):
```
Comment (svlandeg) [inline spacy/tests/test_language.py:842]: As a vaguely related comment, I think we discussed before that we should consider having these overfitting tests in the `slow` test suite, and then run the `slow` tests via the bot on top of its weekly runs. If we agree on that, we could already add the marker here, and go through the other ones in a separate PR on `master`.

File: spacy/ty.py [hunk 9]
```diff
@@ -27,6 +27,25 @@ def finish_update(self, sgd: Optimizer) -> None:
         ...
 
 
+@runtime_checkable
+class DistillableComponent(Protocol):
```
Comment (svlandeg) [inline spacy/ty.py:31]: I love these `Protocol`'s so much 🥳

File: website/docs/api/language.mdx [hunk 10]
```diff
@@ -333,6 +333,34 @@ and custom registered functions if needed. See the
...
+| `losses`        | Dictionary to update with the loss, keyed by pipeline component. ~~Optional[Dict[str, float]]~~                                                                                             |
+| `component_cfg` | Optional dictionary of keyword arguments for components, keyed by component names. Defaults to `None`. ~~Optional[Dict[str, Dict[str, Any]]]~~                                              |
+| `exclude`       | Names of components that shouldn't be updated. Defaults to `[]`. ~~Iterable[str]~~                                                                                                          |
+| `annotates`     | Names of components that should set annotations on the prediced examples after updating. Defaults to `[]`. ~~Iterable[str]~~                                                                |
+| `component_map` | Map student component names to teacher component names, only necessary when the names differ. Defaults to `None`. ~~Iterable[str]~~                                                         |
```
Comment (svlandeg) [inline website/docs/api/language.mdx:361]:
```suggestion
| `component_map` | Map student component names to teacher component names, only necessary when the names differ. Defaults to `None`. ~~Optional[Dict[str, str]]~~                                                         |
```

**Review by svlandeg — COMMENTED** [review]
Looking good! 

Just had a few small remarks/suggestions.

File: spacy/language.py (hunk 4 above)
Comment (danieldk) [reply spacy/language.py]: Done.

File: spacy/language.py (hunk 5 above)
Comment (danieldk) [reply spacy/language.py]: That's more descriptive, changed.

File: spacy/language.py (hunk 6 above)
Comment (danieldk) [reply spacy/language.py]: Done, also renamed `teacher_pipe_name` to `teacher_name` for consistency.

File: spacy/language.py (hunk 7 above)
Comment (danieldk) [reply spacy/language.py]: ✅

File: spacy/tests/test_language.py (hunk 8 above)
Comment (danieldk) [reply spacy/tests/test_language.py:843]: Good idea! Added the `slow` marker. I'll create an Ora ticket as well.

File: website/docs/api/language.mdx (hunk 10 above)
Comment (danieldk) [reply website/docs/api/language.mdx]: ✅

**Review by svlandeg — APPROVED** [review]
